src/replication.c

```diff
@@ -463,95 +421,70 @@ remove_replica_from_list(spec_t *spec, int iofd) {
 }
 
 /*
- * reads on fd for 'len' and copies to 'data'
- * sets 'errorno' if read returns < 0
- * closes fd if read returns < 0 && errno != EAGAIN, and sets fd_closed
- * also closes fd if read return 0, i.e., EOF
- * returns number of bytes read
+ * perform read/write on fd for 'len' according to state
+ * sets 'errorno' if read/write operation returns < 0
+ * closes fd if operation returns < 0 && errno != EAGAIN|EWOULDBLOCK|EINTR,
+ * and sets fd_closed also closes fd if read return 0, i.e., EOF
+ * returns number of bytes read/written
  */
-int64_t
-read_data(int fd, uint8_t *data, uint64_t len, int *errorno, int *fd_closed) {
-	int64_t rc;
+static int64_t
+perform_read_write_on_fd(int fd, uint8_t *data, uint64_t len, int *errorno,
+    int *fd_closed, int state)
+{
+	int64_t rc = -1;
 	uint64_t nbytes = 0;
+
 	if (fd_closed != NULL)
 		*fd_closed = 0;
 	if (errorno != NULL)
 		*errorno = 0;
 
 	while(1) {
-		rc = read(fd, data + nbytes, len - nbytes);
+		switch (state) {
+			case READ_IO_RESP_HDR:
+			case READ_IO_RESP_DATA:
+				rc = read(fd, data + nbytes, len - nbytes);
+				break;
+
+			case WRITE_IO_SEND_HDR:
+			case WRITE_IO_SEND_DATA:
+				rc = write(fd, data + nbytes, len - nbytes);
+				break;
+
+			default:
+				REPLICA_ERRLOG("received invalid state(%d)\n", state);
+				errno = EINVAL;
+				break;
+		}
+
 		if(rc < 0) {
 			if (errorno != NULL)
 				*errorno = errno;
-			if(errno == EAGAIN || errno == EWOULDBLOCK)
+
+			if (errno == EAGAIN || errno == EWOULDBLOCK ||
+			    errno == EINTR) {
 				return nbytes;
-			else {
+			} else {
 				REPLICA_ERRLOG("received err %d on fd %d, closing it..\n", errno, fd);
-				/*
-				 * TODO: cleanup of replica need to be happen
-				 */
 				close(fd);
 				if (fd_closed != NULL)
 					*fd_closed = 1;
 				return -1;
 			}
-		}
-		else if (rc == 0) {
+		} else if (rc == 0) {
```

Comment:
'write' call returning 0 doesn't mean error.. so, that need to be handled differently..